index.js

```diff
@@ -1,39 +1,58 @@
 import {PassThrough as PassThroughStream} from 'node:stream';
+import {finished} from 'node:stream/promises';
 
 export default function mergeStreams(streams) {
 	if (!Array.isArray(streams)) {
 		throw new TypeError(`Expected an array, got \`${typeof streams}\`.`);
 	}
 
-	const passThroughStream = new PassThroughStream({objectMode: true});
+	for (const stream of streams) {
+		if (!(typeof stream?.pipe === 'function')) {
+			throw new TypeError(`Expected a readable stream, got: \`${typeof stream}\`.`);
+		}
+	}
+
+	const objectMode = streams.some(({readableObjectMode}) => readableObjectMode);
+	const readableHighWaterMark = getReadableHighWaterMark(streams, objectMode);
+	const passThroughStream = new PassThroughStream({objectMode, writableHighWaterMark: 0, readableHighWaterMark});
 	passThroughStream.setMaxListeners(Number.POSITIVE_INFINITY);
 
 	if (streams.length === 0) {
 		passThroughStream.end();
 		return passThroughStream;
 	}
 
-	let activeStreams = streams.length;
+	const streamPromises = Promise.all(streams.map(stream => finished(stream, {cleanup: true, readable: true, writable: false})));
 
 	for (const stream of streams) {
-		if (!(typeof stream?.pipe === 'function')) {
-			throw new TypeError(`Expected a stream, got: \`${typeof stream}\`.`);
-		}
+		pipeStream(stream, passThroughStream, streamPromises);
+	}
+
+	return passThroughStream;
+}
 
-		stream.pipe(passThroughStream, {end: false});
+const getReadableHighWaterMark = (streams, objectMode) => {
+	if (streams.length === 0) {
+		return 0;
+	}
 
-		stream.on('end', () => {
-			activeStreams--;
+	const highWaterMarks = streams
+		.filter(({readableObjectMode}) => readableObjectMode === objectMode)
+		.map(({readableHighWaterMark}) => readableHighWaterMark);
+	return Math.max(...highWaterMarks);
+};
 
-			if (activeStreams === 0) {
-				passThroughStream.end();
-			}
-		});
+const pipeStream = async (stream, passThroughStream, streamPromises) => {
+	stream.pipe(passThroughStream, {end: false});
 
-		stream.on('error', error => {
-			passThroughStream.emit('error', error);
-		});
+	try {
+		await streamPromises;
+		passThroughStream.end();
+	} catch (error) {
+		if (error?.code === 'ERR_STREAM_PREMATURE_CLOSE') {
```

Comment:
That's the error thrown by `finished()` on aborts (e.g. due to `.destroy()`).

https://github.com/nodejs/node/blob/e57a32a03a9682857ed83c38447291b3e83b983e/lib/internal/streams/end-of-stream.js#L154